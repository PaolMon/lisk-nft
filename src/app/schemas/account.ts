export const accountSchema = {
    type: "object",
    required: ["ownNFTs"],
    properties: {
      ownNFTs: {
        type: "array",
        fieldNumber: 4,
        items: {
          dataType: "bytes",
        },
      },
    },
    default: {
      ownNFTs: [],
    },
  };

export type AccountType = {
    nft: {
        ownNFTs: Buffer[]
    }
}