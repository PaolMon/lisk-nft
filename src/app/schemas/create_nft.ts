export const createNFTSchema = {
    $id: "lisk/nft/create",
    type: "object",
    required: ["minPurchaseMargin", "initValue", "name"],
    properties: {
        minPurchaseMargin: {
            dataType: "uint32",
            fieldNumber: 1,
        },
        initValue: {
            dataType: "uint64",
            fieldNumber: 2,
        },
        name: {
            dataType: "string",
            fieldNumber: 3,
        },
    },
}

export type createNFT = {
    minPurchaseMargin : number,
    initValue: number,
    name: string
}