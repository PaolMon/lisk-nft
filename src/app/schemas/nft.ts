export const nfts_schema = {
    $id: "lisk/nft/registeredTokens",
    type: "object",
    required: ["registeredNFTTokens"],
    properties: {
        registeredNFTTokens: {
        type: "array",
        fieldNumber: 1,
        items: {
            type: "object",
            required: ["id", "value", "ownerAddress", "minPurchaseMargin", "name"],
            properties: {
            id: {
                dataType: "bytes",
                fieldNumber: 1,
            },
            value: {
                dataType: "uint64",
                fieldNumber: 2,
            },
            ownerAddress: {
                dataType: "bytes",
                fieldNumber: 3,
            },
            minPurchaseMargin: {
                dataType: "uint32",
                fieldNumber: 4,
            },
            name: {
                dataType: "string",
                fieldNumber: 5,
            },
            },
        },
        },
    },
}

export const nft_history_schema = { 
    $id: 'nft/encoded/nftHistory',
    type: 'object',
    required: ['nftHistory'],
    properties: {
    nftHistory: {
        type: 'array',
        fieldNumber: 1,
        items: {
            dataType: 'bytes',
        },
    },
    },
};

export type NFT = {
    id: Buffer,
    value: number,
    ownerAddress: Buffer,
    minPurchaseMargin: number,
    name: string
}

export type NFTS = {
    registeredNFTTokens: NFT[]
}

export type NFT_HISTORY = {
    nftHistory : Buffer[]
}