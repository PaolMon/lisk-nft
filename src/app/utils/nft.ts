import { codec, cryptography } from "lisk-sdk" 
import { nfts_schema, NFT, NFTS } from '../schemas/nft'

const CHAIN_STATE_NFT_TOKENS = "nft:registeredNFTTokens";

export const createNFTToken = ({ name, ownerAddress, nonce, value, minPurchaseMargin }):NFT => {
    const nonceBuffer = Buffer.alloc(8);
    nonceBuffer.writeBigInt64LE(nonce);
    // Create a unique seed by using a combination of the owner account address and the current nonce of the account.
    const seed = Buffer.concat([ownerAddress, nonceBuffer]);
    const id = cryptography.hash(seed);

    return {
        id,
        minPurchaseMargin,
        name,
        ownerAddress,
        value,
    };
};

export const getAllNFTTokens = async (stateStore): Promise<NFT[]> => {
    const registeredTokensBuffer = await stateStore.chain.get(
        CHAIN_STATE_NFT_TOKENS
    );
    if (!registeredTokensBuffer) {
        return [];
    }

    const registeredTokens = codec.decode<NFTS>(
        nfts_schema,
        registeredTokensBuffer
    );

    return registeredTokens.registeredNFTTokens;
};

export const getAllNFTTokensAsJSON = async (dataAccess): Promise<NFT[]> => {
    const registeredTokensBuffer = await dataAccess.getChainState(
        CHAIN_STATE_NFT_TOKENS
    );

    if (!registeredTokensBuffer) {
        return [];
    }

    const registeredTokens = codec.decode<NFTS>(
        nfts_schema,
        registeredTokensBuffer
    );

    return codec.toJSON<NFTS>(nfts_schema, registeredTokens)
        .registeredNFTTokens;
};
  
export const setAllNFTTokens = async (stateStore, NFTTokens) => {
    const registeredTokens = {
        registeredNFTTokens: NFTTokens.sort((a, b) => a.id.compare(b.id)),
    };

    await stateStore.chain.set(
        CHAIN_STATE_NFT_TOKENS,
        codec.encode(nfts_schema, registeredTokens)
    );
};