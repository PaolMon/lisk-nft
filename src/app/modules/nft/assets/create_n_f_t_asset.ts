import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { createNFTSchema, createNFT } from '../../../schemas/create_nft'
import { createNFTToken, getAllNFTTokens, setAllNFTTokens } from '../../../utils/nft'
import { AccountType } from '../../../schemas/account'

export class CreateNFTAsset extends BaseAsset {
	public name = 'createNFT';
	public id = 0;

	// Define schema for asset
	public schema = createNFTSchema

	public validate({ asset }: ValidateAssetContext<createNFT>): void {
		// Validate your asset
		if (asset.initValue <= 0) { 
			throw new Error("NFT init value is too low.");
		} else if (asset.minPurchaseMargin < 0 || asset.minPurchaseMargin > 100) { 
			throw new Error("The NFT minimum purchase value needs to be between 0-100.");
		}
		const regex = /\\|\.|\+|\*|\?|\[|\^|\]|\$|\(|\)|\{|\}|\=|\!|\<|\>|\||\:|\-|\//;
		if (regex.exec(asset.name) !== null) {
			throw new Error("The NFT name must not contain special characters.");
		}
	}

		// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, reducerHandler, stateStore }: ApplyAssetContext<createNFT>): Promise<void> {
		// verify if purchasing nft exists 
		const nftTokens = await getAllNFTTokens(stateStore);
		const nftTokenIndex = nftTokens.findIndex((t) => (t.name == asset.name));
		if (nftTokenIndex >= 0) {
			throw new Error("The NFT name is already registered.")
			//return Promise.reject(new Error("The NFT name is already registered."))
		}
		const senderAddress = transaction.senderAddress;
		const senderAccount = await stateStore.account.get<AccountType>(senderAddress);
		const nftToken = createNFTToken({
			name: asset.name,
			ownerAddress: senderAddress,
			nonce: transaction.nonce,
			value: asset.initValue,
			minPurchaseMargin: asset.minPurchaseMargin,
		});
	
		// update sender account with unique NFT ID 
		senderAccount.nft.ownNFTs.push(nftToken.id);
		await stateStore.account.set<AccountType>(senderAddress, senderAccount);
	
		// debit tokens from sender account to create an NFT 
		await reducerHandler.invoke("token:debit", {
			address: senderAddress,
			amount: asset.initValue,
		});
	
		// save NFTs 
		const allTokens = await getAllNFTTokens(stateStore);
		allTokens.push(nftToken);
		await setAllNFTTokens(stateStore, allTokens);
	}
	
}
