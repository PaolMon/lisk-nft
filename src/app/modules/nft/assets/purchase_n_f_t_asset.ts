import { BaseAsset, ApplyAssetContext } from 'lisk-sdk';
import { getAllNFTTokens, setAllNFTTokens } from '../../../utils/nft'
import { purchaseNFTSchema, PurchaseNFT } from '../../../schemas/purchase_nft'
import { AccountType } from '../../../schemas/account'


export class PurchaseNFTAsset extends BaseAsset {
	public name = 'purchaseNFT';
  	public id = 1;

  	// Define schema for asset
	public schema = purchaseNFTSchema;

	// eslint-disable-next-line @typescript-eslint/require-await
	public async apply({ asset, transaction, reducerHandler, stateStore }: ApplyAssetContext<PurchaseNFT>): Promise<void> {
		// verify if purchasing nft exists 
		const nftTokens = await getAllNFTTokens(stateStore);
		const nftTokenIndex = nftTokens.findIndex((t) => t.id.equals(asset.nftId));
	
		if (nftTokenIndex < 0) {
			throw new Error("Token id not found");
		}
		// verify if minimum nft purchasing condition met 
		const token = nftTokens[nftTokenIndex];
		const tokenOwner = await stateStore.account.get<AccountType>(token.ownerAddress);
		const tokenOwnerAddress = tokenOwner.address;
	
		if (token && token.minPurchaseMargin === 0) {
			throw new Error("This NFT can not be purchased");
		}
	
		const tokenCurrentValue = BigInt(token.value);
		const tokenMinPurchaseValue =
		tokenCurrentValue +
		(tokenCurrentValue * BigInt(token.minPurchaseMargin)) / BigInt(100);
		const purchaseValue = asset.purchaseValue;
	
		if (tokenMinPurchaseValue > purchaseValue) {
			throw new Error("Token can not be purchased. Purchase value is too low.");
		}
	
		// remove nft from owner account 
		const purchaserAddress = transaction.senderAddress;
		const purchaserAccount = await stateStore.account.get<AccountType>(purchaserAddress);
	
		const ownerTokenIndex = tokenOwner.nft.ownNFTs.findIndex((a) =>
			a.equals(token.id)
		);
		tokenOwner.nft.ownNFTs.splice(ownerTokenIndex, 1);
		await stateStore.account.set<AccountType>(tokenOwnerAddress, tokenOwner);
	
		// add nft to purchaser account 
		purchaserAccount.nft.ownNFTs.push(token.id);
		await stateStore.account.set<AccountType>(purchaserAddress, purchaserAccount);
	
		token.ownerAddress = purchaserAddress;
		token.value = purchaseValue;
		nftTokens[nftTokenIndex] = token;
		await setAllNFTTokens(stateStore, nftTokens);
	
		// debit LSK tokens from purchaser account 
		await reducerHandler.invoke("token:debit", {
			address: purchaserAddress,
			amount: purchaseValue,
		});
	
		// credit LSK tokens to purchaser account 
		await reducerHandler.invoke("token:credit", {
			address: tokenOwnerAddress,
			amount: purchaseValue,
		});
	}
}

