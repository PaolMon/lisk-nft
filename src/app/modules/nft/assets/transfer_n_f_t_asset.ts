import { BaseAsset, ApplyAssetContext } from 'lisk-sdk';
import { transferNFTSchema, TransferNFT } from '../../../schemas/transfer_nft'
import { getAllNFTTokens, setAllNFTTokens } from '../../../utils/nft'
import { AccountType } from '../../../schemas/account'

export class TransferNFTAsset extends BaseAsset {
	public name = 'transferNFT';
  	public id = 2;

  // Define schema for asset
	public schema = transferNFTSchema;

  	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<TransferNFT>): Promise<void> {
		const nftTokens = await getAllNFTTokens(stateStore);
		const nftTokenIndex = nftTokens.findIndex((t) => t.id.equals(asset.nftId));

		// 4.verify if the nft exists 
		if (nftTokenIndex < 0) {
			throw new Error("Token id not found");
		}
		const token = nftTokens[nftTokenIndex];
		const tokenOwnerAddress = token.ownerAddress;
		const senderAddress = transaction.senderAddress;
		// 5.verify that the sender owns the nft 

		if (!tokenOwnerAddress.equals(senderAddress)) {
			throw new Error("An NFT can only be transferred by the owner of the NFT.");
		}

		const tokenOwner = await stateStore.account.get<AccountType>(tokenOwnerAddress);
		// 6.remove nft from the owner account 
		const ownerTokenIndex = tokenOwner.nft.ownNFTs.findIndex((a) =>
			a.equals(token.id)
		);
		tokenOwner.nft.ownNFTs.splice(ownerTokenIndex, 1);
		await stateStore.account.set(tokenOwnerAddress, tokenOwner);

		// 7.add nft to the recipient account 
		const recipientAddress = asset.recipient;
		const recipientAccount = await stateStore.account.get<AccountType>(recipientAddress);
		recipientAccount.nft.ownNFTs.push(token.id);
		await stateStore.account.set(recipientAddress, recipientAccount);

		token.ownerAddress = recipientAddress;
		nftTokens[nftTokenIndex] = token;
		await setAllNFTTokens(stateStore, nftTokens);
	}
}
