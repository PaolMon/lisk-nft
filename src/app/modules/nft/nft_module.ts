/*
 * LiskHQ/lisk-commander
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

/* eslint-disable class-methods-use-this */

import {
    AfterBlockApplyContext,


    AfterGenesisBlockApplyContext, BaseModule,


    BeforeBlockApplyContext, TransactionApplyContext
} from 'lisk-sdk';
import { CreateNFTAsset } from "./assets/create_n_f_t_asset";
import { PurchaseNFTAsset } from "./assets/purchase_n_f_t_asset";
import { TransferNFTAsset } from "./assets/transfer_n_f_t_asset";
import { accountSchema } from '../../schemas/account'
import { getAllNFTTokensAsJSON } from '../../utils/nft';

export class NftModule extends BaseModule {
    public accountSchema = accountSchema;
    public actions = {
        // Example below
        // getBalance: async (params) => this._dataAccess.account.get(params.address).token.balance,
        // getBlockByID: async (params) => this._dataAccess.blocks.get(params.id),
        getAllNFTTokens: async () => { console.log('\nGET ALL NFT TOKEN ACTION CALLED \n') ; getAllNFTTokensAsJSON(this._dataAccess) },
    };
    public reducers = {
        // Example below
        // getBalance: async (
		// 	params: Record<string, unknown>,
		// 	stateStore: StateStore,
		// ): Promise<bigint> => {
		// 	const { address } = params;
		// 	if (!Buffer.isBuffer(address)) {
		// 		throw new Error('Address must be a buffer');
		// 	}
		// 	const account = await stateStore.account.getOrDefault<TokenAccount>(address);
		// 	return account.token.balance;
		// },
    };
    public name = 'nft';
    public transactionAssets = [new CreateNFTAsset(), new PurchaseNFTAsset(), new TransferNFTAsset()];
    public events = [
        // Example below
        // 'nft:newBlock',
    ];
    public id = 1024;

    // public constructor(genesisConfig: GenesisConfig) {
    //     super(genesisConfig);
    // }

    // Lifecycle hooks
    public async beforeBlockApply(_input: BeforeBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async afterBlockApply(_input: AfterBlockApplyContext) {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
		// const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async beforeTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext) {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext) {
        // Get any data from genesis block, for example get all genesis accounts
        // const genesisAccoounts = genesisBlock.header.asset.accounts;
    }
}
