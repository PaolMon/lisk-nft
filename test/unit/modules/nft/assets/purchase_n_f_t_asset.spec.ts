import { PurchaseNFTAsset } from '../../../../../src/app/modules/nft/assets/purchase_n_f_t_asset';
import { StateStore, ReducerHandler,testing } from 'lisk-sdk';
import { AccountType } from '../../../../../src/app/schemas/account';
import { NftModule } from '../../../../../src/app/modules/nft/nft_module';
import { getAllNFTTokens } from '../../../../../src/app/utils/nft';
import { NFT } from '../../../../../src/app/schemas/nft';
import { CreateNFTAsset } from '../../../../../src/app/modules/nft/assets/create_n_f_t_asset';

describe('PurchaseNFTAsset', () => {
  let transactionAsset: PurchaseNFTAsset;
  let createAsset: CreateNFTAsset;
  let stateStore: StateStore;
  let reducerHandler: ReducerHandler;
  let account: any;
  let creatorAccount: any;
  //let nftToken: NFT;
  let context;
  let creatContext;

  beforeEach(async () => {
	// Create new account
	account = testing.fixtures.createDefaultAccount<AccountType>([NftModule]);
	creatorAccount = testing.fixtures.createDefaultAccount<AccountType>([NftModule]);

	transactionAsset = new PurchaseNFTAsset();
	createAsset = new CreateNFTAsset();

	// Create state store mock with account
	stateStore = new testing.mocks.StateStoreMock({
		accounts: [account, creatorAccount],
	});

	// Create reducer handler mock
	reducerHandler = testing.mocks.reducerHandlerMock;

	// creatContextContext used to create a new NFT used by tests
	creatContext = testing.createApplyAssetContext({
		stateStore,
		reducerHandler,
		asset: { name: 'LOKI', initValue: BigInt(10), minPurchaseMargin: 10 },
		transaction: { senderAddress: creatorAccount.address, nonce: BigInt(2) } as any,
	});

	await createAsset.apply(creatContext);

	// Tracks calls to stateStore.chain and the reducerHandler
	jest.spyOn(stateStore.chain, 'get');
	jest.spyOn(stateStore.chain, 'set');
	jest.spyOn(reducerHandler, 'invoke');

});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(1);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('purchaseNFT');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});

	describe('validate', () => {
		describe('schema validation', () => {
			it.todo('should throw errors for invalid schema');
			it.todo('should be ok for valid schema');
		});
	});

	describe('apply', () => {
		describe('valid cases', () => {
			it.todo('should update the state store');
			it('should update sender account with nft\' s id', async () => {
				const nftTokens = await getAllNFTTokens(stateStore);
				const nftTokenIndex = nftTokens.findIndex((t) => (t.name == 'LOKI'));
				const purchaseToken: NFT = nftTokens[nftTokenIndex];
				// Create context for the apply() function
				context = testing.createApplyAssetContext({
					stateStore,
					reducerHandler,
					asset: { name: purchaseToken.name, nftId: purchaseToken.id, purchaseValue: 20 },
					transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
				});
				await transactionAsset.apply(context);
				const purchaser = await stateStore.account.get<AccountType>(account.address);
				expect(purchaser.nft.ownNFTs).toContainEqual(purchaseToken.id)
			})
		});

		describe('invalid cases', () => {
			it.todo('should throw error');
			it('should throw error because purchase value is too low', async () => {
				const nftTokens = await getAllNFTTokens(stateStore);
				const nftTokenIndex = nftTokens.findIndex((t) => (t.name == 'LOKI'));
				const purchaseToken: NFT = nftTokens[nftTokenIndex];
				// Create context for the apply() function
				context = testing.createApplyAssetContext({
					stateStore,
					reducerHandler,
					asset: { name: purchaseToken.name, nftId: purchaseToken.id, purchaseValue: 1 },
					transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
				});
				await expect(transactionAsset.apply(context)).rejects.toThrow("Token can not be purchased. Purchase value is too low.");
			})
		});
	});
});
