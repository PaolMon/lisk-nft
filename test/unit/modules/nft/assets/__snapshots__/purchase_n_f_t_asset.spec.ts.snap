// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PurchaseNFTAsset constructor should have valid schema 1`] = `
Object {
  "$id": "lisk/nft/purchase",
  "properties": Object {
    "name": Object {
      "dataType": "string",
      "fieldNumber": 3,
    },
    "nftId": Object {
      "dataType": "bytes",
      "fieldNumber": 1,
    },
    "purchaseValue": Object {
      "dataType": "uint64",
      "fieldNumber": 2,
    },
  },
  "required": Array [
    "nftId",
    "purchaseValue",
    "name",
  ],
  "type": "object",
}
`;
